vyper/builtin_functions/convert.py

```diff
@@ -1,528 +1,387 @@
-                    return LLLnode.from_list(
-                        ["mul", in_arg, DECIMAL_DIVISOR], typ=BaseType("decimal"), pos=getpos(expr)
-                    )
-            else:
-                return LLLnode.from_list(
-                    [
-                        "uclample",
-                        ["mul", in_arg, DECIMAL_DIVISOR],
-                        ["mload", MemoryPositions.MAXDECIMAL],
-                    ],
-                    typ=BaseType("decimal"),
-                    pos=getpos(expr),
-                )
-
-        elif input_type == "address":
-            return LLLnode.from_list(
-                [
-                    "mul",
-                    ["signextend", 15, ["and", in_arg, (SizeLimits.ADDRSIZE - 1)]],
-                    DECIMAL_DIVISOR,
-                ],
-                typ=BaseType("decimal"),
-                pos=getpos(expr),
-            )
-
-        elif input_type == "bytes32":
-            if in_arg.typ.is_literal:
-                if not SizeLimits.in_bounds("int128", (in_arg.value * DECIMAL_DIVISOR)):
-                    raise InvalidLiteral(
-                        f"Number out of range: {in_arg.value}",
-                        expr,
-                    )
-                else:
-                    return LLLnode.from_list(
-                        ["mul", in_arg, DECIMAL_DIVISOR], typ=BaseType("decimal"), pos=getpos(expr)
-                    )
-            else:
-                return LLLnode.from_list(
-                    [
-                        "clamp",
-                        ["mload", MemoryPositions.MINDECIMAL],
-                        ["mul", in_arg, DECIMAL_DIVISOR],
-                        ["mload", MemoryPositions.MAXDECIMAL],
-                    ],
-                    typ=BaseType("decimal"),
-                    pos=getpos(expr),
-                )
-
-        elif input_type == "int256":
-            # cast in_arg so clamp_basetype works
-            in_arg = LLLnode.from_list(in_arg, typ="int128")
-            return LLLnode.from_list(
-                ["mul", clamp_basetype(in_arg), DECIMAL_DIVISOR],
-                typ=BaseType("decimal"),
-                pos=getpos(expr),
-            )
-
-        elif input_type in ("uint8", "int128", "bool"):
-            return LLLnode.from_list(
-                ["mul", in_arg, DECIMAL_DIVISOR], typ=BaseType("decimal"), pos=getpos(expr)
-            )
+        val = decimal.Decimal(expr.value)  # should work for Int, Decimal, Hex
+
+    val = val * DECIMAL_DIVISOR
+
+    (lo, hi) = (SizeLimits.MINDECIMAL, SizeLimits.MAXDECIMAL)
+    if not (lo <= val <= hi):
+        raise InvalidLiteral("Number out of range", expr)
+
+    # sanity check type checker did its job
+    assert math.ceil(val) == math.floor(val)
+
+    return LLLnode.from_list(int(val), typ=out_typ)
+
+
+# any base type or bytes/string
+@_input_types("int", "decimal", "bytes_m", "address", "bool", "bytes", "string")
+def to_bool(expr, arg, out_typ):
```

Comment:
nit: document public functions, also why is this one public but some of the other conversion functions aren't?